ipa-core/src/protocol/hybrid/oprf.rs

```diff
@@ -0,0 +1,317 @@
+            validate: &HybridStep::ConvertFp25519Validate,
+        },
+        CONV_PROOF_CHUNK,
+    );
+    let m_ctx = validator.context();
+
+    let curve_pts = seq_join(
+        ctx.active_work(),
+        process_slice_by_chunks(input_rows, move |idx, records: ChunkData<_, CONV_CHUNK>| {
+            let record_id = RecordId::from(idx);
+            let input_match_keys: &dyn Fn(usize) -> Replicated<MatchKey> =
+                &|i| records[i].match_key.clone();
+            let match_keys =
+                BitDecomposed::<Replicated<Boolean, 256>>::transposed_from(input_match_keys)
+                    .unwrap_infallible();
+            convert_to_fp25519::<_, CONV_CHUNK, PRF_CHUNK>(m_ctx.clone(), record_id, match_keys)
+        }),
+    )
+    .map_ok(Chunk::unpack::<PRF_CHUNK>)
+    .try_flatten_iters()
+    .try_collect::<Vec<_>>()
+    .await?;
+
+    // In order to compute the same PRF value for the same match key
+    // all shards must have the same prf_key. This has the leader
+    // generate a key, and give it to all the followers.
+    let prf_key_ctx = ctx.set_total_records(TotalRecords::ONE);
+    let prf_key = if ctx.shard_id() == ShardIndex::FIRST {
+        let leader_prf_key = gen_prf_key(&ctx.narrow(&HybridStep::PrfKeyGen));
+
+        for shard in ctx.shard_count().iter().skip(1) {
+            prf_key_ctx
+                .clone()
+                .shard_send_channel::<Replicated<Fp25519>>(shard)
+                .send(RecordId::FIRST, leader_prf_key.clone())
+                .await?;
+        }
+        Ok(leader_prf_key)
+    } else {
+        let mut receiver = prf_key_ctx
+            .shard_recv_channel::<Replicated<Fp25519>>(ShardIndex::FIRST)
+            .take(1);
+        receiver.next().await.unwrap()
+    }
+    .unwrap();
+
+    let validator = ctx
+        .narrow(&HybridStep::EvalPrf)
+        .set_total_records(eval_records)
+        .validator::<Fp25519>();
+    let eval_ctx = validator.context();
+
+    let prf_of_match_keys = seq_join(
+        ctx.active_work(),
+        stream::iter(curve_pts).enumerate().map(|(i, curve_pts)| {
+            let record_id = RecordId::from(i);
+            let eval_ctx = eval_ctx.clone();
+            let prf_key = &prf_key;
+            curve_pts
+                .then(move |pts| eval_dy_prf::<_, PRF_CHUNK>(eval_ctx, record_id, prf_key, pts))
+        }),
+    )
+    .try_collect::<Vec<_>>()
+    .await?;
+
+    Ok(zip(input_rows, prf_of_match_keys.into_iter().flatten())
+        .map(|(input, prf_of_match_key)| {
+            let IndistinguishableHybridReport {
+                match_key: _,
+                value,
+                breakdown_key,
+            } = &input;
+
+            PRFIndistinguishableHybridReport {
+                prf_of_match_key,
+                value: value.clone(),
+                breakdown_key: breakdown_key.clone(),
+            }
+        })
+        .collect())
+}
+
+#[cfg(all(test, not(feature = "shuttle")))]
```

Comment:
in order to get the tests to work with `multi-threading`, I needed to use `test_executor::run`, and it seems that's [incompatible with the `shuttle` feature](https://github.com/private-attribution/ipa/blob/main/ipa-core/src/lib.rs#L292). 

can someone confirm this is ok? I don't want to ignore a real issue.

Reply:
Don't think run is incompatible. If you scroll up on that link you will see there's another implementation for run with shuttle.

Here are some tests you can check that also use run
https://github.com/private-attribution/ipa/blob/main/ipa-core/src/query/processor.rs#L623

Reply:
thanks for the code pointer. let me see if I can get this running. it may be failing for a different reason